File: apps/mobile/modules/nemu-aidoku/runtime/kotlin/NemuNativeHttpAddressPolicy.kt
package pm.nemu.mobile.aidoku

import java.io.IOException
import java.net.InetAddress
import java.net.Proxy
import java.net.UnknownHostException
import java.util.Locale
import okhttp3.Dns
import okhttp3.Interceptor
import okhttp3.Response

/**
 * Network boundary for untrusted source packages.
 *
 * A source may choose arbitrary request URLs, so accepting only HTTP(S) is not
 * sufficient: without an address policy it can read loopback/LAN/cloud-metadata
 * services and exfiltrate the response in a later public request. DNS removes
 * every non-public answer and requires at least one public route; OkHttp can
 * therefore select only a validated address. The network interceptor re-checks
 * the connected peer before request bytes are written, which also closes
 * DNS-rebinding and redirect races after lookup.
 */
internal object NemuNativeHttpAddressPolicy {
  private const val BLOCKED_DESTINATION_MESSAGE =
    "Native source networking blocked a private or reserved destination."

  internal fun resolvePublicAddresses(
    hostname: String,
    resolver: (String) -> List<InetAddress>
  ): List<InetAddress> {
    val normalized = normalizeHostname(hostname)
    if (isForbiddenHostname(normalized)) throw blockedDestination()

    val addresses = try {
      resolver(normalized)
    } catch (error: UnknownHostException) {
      throw error
    } catch (error: Throwable) {
      throw UnknownHostException("Native source host lookup failed.").also {
        it.initCause(error)
      }
    }
    if (addresses.isEmpty()) {
      throw UnknownHostException("Native source host lookup returned no addresses.")
    }
    val validatedAddresses = addresses.filter {
      isPublicAddress(it.address) ||
        (!isNumericHostname(normalized) && isProxySyntheticAddress(it.address))
    }
    if (validatedAddresses.isEmpty()) throw blockedDestination()
    return validatedAddresses
  }

  internal fun requirePublicAddress(address: InetAddress, hostname: String) {
    val normalized = normalizeHostname(hostname)
    val allowed = isPublicAddress(address.address) ||
      (!isNumericHostname(normalized) && isProxySyntheticAddress(address.address))
    if (!allowed) throw blockedDestination()
  }

  /**
   * Validates a destination before OkHttp is allowed to build a route for it.
   *
   * OkHttp resolves IP-literal hosts itself and never calls [Dns.lookup] for
   * them, so [NemuPublicAddressDns] cannot see `http://127.0.0.1/` at all. The
   * only remaining guard is a network interceptor, which runs after the TCP
   * connect and TLS handshake have already happened. Check literals here, in
   * the application layer, so a private destination fails closed before any
   * connection is attempted.
   */
  internal fun requirePublicDestination(hostname: String) {
    val normalized = normalizeHostname(hostname)
    if (isForbiddenHostname(normalized)) throw blockedDestination()
    if (!isNumericHostname(normalized)) return
    val literal = parseIpLiteral(normalized) ?: throw blockedDestination()
    if (!isPublicAddress(literal)) throw blockedDestination()
  }

  /**
   * A host that can only be an address literal. Anything numeric that is not a
   * strict literal is rejected rather than resolved: no registry name is fully
   * numeric, and platform resolvers have historically accepted octal, decimal
   * and short-form spellings of loopback and link-local addresses.
   */
  internal fun isNumericHostname(hostname: String): Boolean {
    if (hostname.isEmpty()) return false
    if (hostname.contains(':')) return true
    return hostname.all { it in '0'..'9' || it == '.' }
  }

  /**
   * Surge and compatible TUN proxies synthesize DNS answers from RFC 2544's
   * benchmarking range. The range remains blocked as a direct URL literal and
   * is accepted only when associated with an ordinary DNS hostname.
   */
  internal fun isProxySyntheticAddress(bytes: ByteArray): Boolean {
    return bytes.size == 4 &&
      unsigned(bytes[0]) == 198 &&
      unsigned(bytes[1]) in 18..19
  }

  /** Address bytes for an IP literal, or `null` when [value] is not one. */
  internal fun parseIpLiteral(value: String): ByteArray? {
    if (value.isEmpty()) return null
    return if (value.contains(':')) parseIpv6Literal(value) else parseIpv4Literal(value)
  }

  private fun parseIpv4Literal(value: String): ByteArray? {
    val parts = value.split('.')
    if (parts.size != 4) return null
    val bytes = ByteArray(4)
    parts.forEachIndexed { index, part ->
      if (part.isEmpty() || part.length > 3) return null
      if (!part.all { it in '0'..'9' }) return null
      val octet = part.toInt()
      if (octet > 255) return null
      bytes[index] = octet.toByte()
    }
    return bytes
  }

  private fun parseIpv6Literal(value: String): ByteArray? {
    // A scoped literal only ever names a local interface.
    if (value.contains('%')) return null
    val compression = value.indexOf("::")
    if (compression >= 0 && compression != value.lastIndexOf("::")) return null
    val headText = if (compression >= 0) value.substring(0, compression) else value
    val tailText = if (compression >= 0) value.substring(compression + 2) else ""
    if (headText.endsWith(':') || tailText.startsWith(':') || tailText.endsWith(':')) {
      return null
    }
    if (compression < 0 && headText.startsWith(':')) return null

    val headGroups = if (headText.isEmpty()) emptyList() else headText.split(':')
    val tailGroups = if (tailText.isEmpty()) emptyList() else tailText.split(':')
    val head = ipv6GroupBytes(
      headGroups,
      allowEmbeddedIpv4 = compression < 0
    ) ?: return null
    val tail = ipv6GroupBytes(
      tailGroups,
      allowEmbeddedIpv4 = true
    ) ?: return null

    if (compression < 0) return if (head.size == 16) head else null
    val zeroes = 16 - head.size - tail.size
    if (zeroes < 1) return null
    return head + ByteArray(zeroes) + tail
  }

  private fun ipv6GroupBytes(groups: List<String>, allowEmbeddedIpv4: Boolean): ByteArray? {
    val output = ArrayList<Byte>(16)
    groups.forEachIndexed { index, group ->
      if (group.contains('.')) {
        if (!allowEmbeddedIpv4 || index != groups.size - 1) return null
        val embedded = parseIpv4Literal(group) ?: return null
        for (byte in embedded) output.add(byte)
        return@forEachIndexed
      }
      if (group.isEmpty() || group.length > 4) return null
      if (!group.all { it in '0'..'9' || it in 'a'..'f' }) return null
      val parsed = group.toInt(16)
      output.add(((parsed shr 8) and 0xff).toByte())
      output.add((parsed and 0xff).toByte())
    }
    if (output.size > 16) return null
    return output.toByteArray()
  }

  internal fun isForbiddenHostname(hostname: String): Boolean {
    val normalized = normalizeHostname(hostname)
    if (normalized.isEmpty()) return true
    if (
      normalized == "localhost" ||
      normalized.endsWith(".localhost") ||
      normalized == "local" ||
      normalized.endsWith(".local") ||
      normalized == "localdomain" ||
      normalized.endsWith(".localdomain") ||
      normalized == "internal" ||
      normalized.endsWith(".internal") ||
      normalized == "home.arpa" ||
      normalized.endsWith(".home.arpa")
    ) {
      return true
    }

    return normalized == "metadata" ||
      normalized == "metadata.goog" ||
      normalized.endsWith(".metadata.goog") ||
      normalized == "instance-data" ||
      normalized == "instance-data.ec2.internal" ||
      normalized == "metadata.aws.internal" ||
      normalized == "metadata.azure.internal"
  }

  internal fun isPublicAddress(bytes: ByteArray): Boolean {
    return when (bytes.size) {
      4 -> isPublicIpv4(bytes, 0)
      16 -> isPublicIpv6(bytes)
      else -> false
    }
  }

  private fun isPublicIpv4(bytes: ByteArray, offset: Int): Boolean {
    val first = unsigned(bytes[offset])
    val second = unsigned(bytes[offset + 1])
    val third = unsigned(bytes[offset + 2])
    val fourth = unsigned(bytes[offset + 3])

    if (first == 0 || first == 10 || first == 127 || first >= 224) return false
    if (first == 100 && second in 64..127) return false // shared/CGNAT
    if (first == 169 && second == 254) return false // link-local + metadata
    if (first == 172 && second in 16..31) return false
    if (
      first == 192 && second == 0 && third == 0 && fourth != 9 && fourth != 10
    ) return false
    if (first == 192 && second == 0 && third == 2) return false
    if (first == 192 && second == 88 && third == 99 && fourth != 2) return false
    if (first == 192 && second == 168) return false
    if (first == 198 && second in 18..19) return false // benchmark networks
    if (first == 198 && second == 51 && third == 100) return false
    if (first == 203 && second == 0 && third == 113) return false
    return true
  }

  private fun isPublicIpv6(bytes: ByteArray): Boolean {
    // IPv4-mapped IPv6. Some JDKs expose these as four-byte Inet4Address values,
    // but handling the 16-byte representation prevents platform drift bypasses.
    if (
      bytes.take(10).all { it.toInt() == 0 } &&
      unsigned(bytes[10]) == 0xff &&
      unsigned(bytes[11]) == 0xff
    ) {
      return isPublicIpv4(bytes, 12)
    }

    // The well-known NAT64 prefix is public only when its embedded IPv4 address
    // is public. The local-use 64:ff9b:1::/48 prefix remains blocked below.
    if (
      unsigned(bytes[0]) == 0x00 &&
      unsigned(bytes[1]) == 0x64 &&
      unsigned(bytes[2]) == 0xff &&
      unsigned(bytes[3]) == 0x9b &&
      bytes.sliceArray(4 until 12).all { it.toInt() == 0 }
    ) {
      return isPublicIpv4(bytes, 12)
    }

    // Only currently allocated global-unicast space is eligible. This rejects
    // unspecified, loopback, link/site-local, ULA, multicast, discard-only and
    // other reserved blocks before the narrower exclusions below.
    if ((unsigned(bytes[0]) and 0xe0) != 0x20) return false // 2000::/3

    // 6to4 embeds an IPv4 destination in bits 16...48.
    if (unsigned(bytes[0]) == 0x20 && unsigned(bytes[1]) == 0x02) {
      return isPublicIpv4(bytes, 2)
    }

    if (
      unsigned(bytes[0]) == 0x3f &&
      unsigned(bytes[1]) == 0xff &&
      (unsigned(bytes[2]) and 0xf0) == 0
    ) {
      return false // 3fff::/20 documentation space
    }
    if (unsigned(bytes[0]) == 0x20 && unsigned(bytes[1]) == 0x01) {
      val third = unsigned(bytes[2])
      val fourth = unsigned(bytes[3])
      if (third == 0x00 && fourth == 0x00) return false // Teredo special range
      if (
        third == 0x00 &&
        fourth == 0x02 &&
        unsigned(bytes[4]) == 0 &&
        unsigned(bytes[5]) == 0
      ) return false // benchmarking
      if (third == 0x0d && fourth == 0xb8) return false // documentation
      if (third == 0x00 && (fourth and 0xf0) in setOf(0x10, 0x20, 0x30)) {
        return false // ORCHID/ORCHIDv2/Drone DETs
      }
    }
    return true
  }

  private fun normalizeHostname(hostname: String): String {
    return hostname
      .trim()
      .removePrefix("[")
      .removeSuffix("]")
      .trimEnd('.')
      .lowercase(Locale.US)
  }

  private fun unsigned(value: Byte): Int = value.toInt() and 0xff

  private fun blockedDestination(): UnknownHostException {
    return UnknownHostException(BLOCKED_DESTINATION_MESSAGE)
  }
}

/** Returns only validated DNS answers to OkHttp, including redirect lookups. */
internal object NemuPublicAddressDns : Dns {
  override fun lookup(hostname: String): List<InetAddress> {
    return NemuNativeHttpAddressPolicy.resolvePublicAddresses(hostname) {
      Dns.SYSTEM.lookup(it)
    }
  }
}

/**
 * Runs before OkHttp selects a route, so an IP-literal destination is refused
 * without opening a socket. Application interceptors do not re-run for redirect
 * hops; those stay covered by [NemuPublicAddressDns] and the connected-peer
 * network interceptor below.
 */
internal class NemuPublicAddressPreflightInterceptor : Interceptor {
  override fun intercept(chain: Interceptor.Chain): Response {
    NemuNativeHttpAddressPolicy.requirePublicDestination(chain.request().url.host)
    return chain.proceed(chain.request())
  }
}

/**
 * Runs once per network hop after connect but before request headers/body. With
 * proxies disabled on the client, the route address is the actual destination;
 * rejecting it here prevents a DNS answer that changed after lookup from being
 * used for SSRF.
 */
internal class NemuPublicAddressNetworkInterceptor : Interceptor {
  override fun intercept(chain: Interceptor.Chain): Response {
    val address = chain.connection()?.route()?.socketAddress?.address
      ?: throw IOException("Native source destination could not be validated.")
    NemuNativeHttpAddressPolicy.requirePublicAddress(
      address,
      chain.request().url.host
    )
    return chain.proceed(chain.request())
  }
}

internal val NEMU_NATIVE_HTTP_DIRECT_PROXY: Proxy = Proxy.NO_PROXY

File: apps/mobile/modules/nemu-aidoku/ios/NemuNativeHttpAddressPolicy.swift
import Darwin
import Foundation

enum NemuNativeHttpAddressPolicyError: Error, LocalizedError {
  case invalidURL
  case blockedDestination
  case unresolvedHost

  var errorDescription: String? {
    switch self {
    case .invalidURL:
      return "Only HTTP and HTTPS source URLs are allowed."
    case .blockedDestination:
      return "Native source networking blocked a private or reserved destination."
    case .unresolvedHost:
      return "Network unavailable or source host could not be resolved safely."
    }
  }
}

/**
 * Address policy for untrusted source-package networking.
 *
 * Every source destination is resolved inside Nemu's authenticated loopback
 * proxy. Private and reserved answers are removed, and at least one public
 * answer must remain. The proxy then opens a socket to one of those exact IP
 * literals, so filtered answers cannot be selected and no second resolver
 * lookup can rebind the connection after validation. URLSession's metrics are
 * retained as a fail-closed assertion that every transaction used that
 * loopback proxy instead of contacting an origin directly.
 */
enum NemuNativeHttpAddressPolicy {
  typealias Resolver = (String) throws -> [[UInt8]]

  private static let validationQueue = DispatchQueue(
    label: "pm.nemu.native-http-address-policy",
    qos: .utility,
    attributes: .concurrent
  )

  static func validatedURL(_ value: String) throws -> URL {
    guard
      let url = URL(string: value),
      let scheme = url.scheme?.lowercased(),
      scheme == "http" || scheme == "https",
      url.host?.isEmpty == false
    else {
      throw NemuNativeHttpAddressPolicyError.invalidURL
    }
    try validate(url: url)
    return url
  }

  /// URL.port omits a scheme's default port. Compare effective ports so a
  /// canonical redirect between `https://host` and `https://host:443` (or
  /// HTTP/80) remains same-origin without weakening cross-port isolation.
  static func effectivePort(for url: URL?) -> Int? {
    guard let url else { return nil }
    if let port = url.port { return port }
    switch url.scheme?.lowercased() {
    case "https": return 443
    case "http": return 80
    default: return nil
    }
  }

  static func validate(url: URL) throws {
    try validate(url: url, resolver: resolveAddresses)
  }

  static func validateAsync(
    url: URL,
    completion: @escaping @Sendable (Bool) -> Void
  ) {
    validationQueue.async {
      do {
        try validate(url: url)
        completion(true)
      } catch {
        completion(false)
      }
    }
  }

  static func validate(url: URL, resolver: Resolver) throws {
    guard let host = url.host else {
      throw NemuNativeHttpAddressPolicyError.invalidURL
    }
    _ = try validatedAddresses(hostname: host, resolver: resolver)
  }

  static func validatedAddresses(hostname: String) throws -> [[UInt8]] {
    try validatedAddresses(hostname: hostname, resolver: resolveAddresses)
  }

  static func validatedAddresses(
    hostname: String,
    resolver: Resolver
  ) throws -> [[UInt8]] {
    let normalized = normalizeHostname(hostname)
    guard !isForbiddenHostname(normalized) else {
      throw NemuNativeHttpAddressPolicyError.blockedDestination
    }

    let addresses: [[UInt8]]
    do {
      addresses = try resolver(normalized)
    } catch let error as NemuNativeHttpAddressPolicyError {
      throw error
    } catch {
      throw NemuNativeHttpAddressPolicyError.unresolvedHost
    }
    guard !addresses.isEmpty else {
      throw NemuNativeHttpAddressPolicyError.unresolvedHost
    }
    let allowProxySyntheticAddresses = !isNumericHostname(normalized)
    let validatedAddresses = addresses.filter { address in
      isPublicAddress(address) ||
        (allowProxySyntheticAddresses && isProxySyntheticAddress(address))
    }
    guard !validatedAddresses.isEmpty else {
      throw NemuNativeHttpAddressPolicyError.blockedDestination
    }
    return validatedAddresses
  }

  /// Fail-closed defense in depth for the explicit proxy configuration. The
  /// pre-send guarantee comes from the loopback proxy connecting to a validated
  /// IP literal; this gate prevents a response from crossing into a source if a
  /// future URLSession behavior change ever bypasses that configured proxy.
  static func validateLoopbackProxy(
    remoteAddress: String?,
    remotePort: Int?,
    isProxyConnection: Bool,
    requestScheme: String?,
    expectedPort: UInt16?
  ) throws {
    guard
      // Legacy CFNetwork reports absolute-form HTTP proxy transactions as
      // non-proxy even though remoteAddress/remotePort identify our listener.
      // HTTPS (including CONNECT) must still carry the explicit proxy bit.
      isProxyConnection || requestScheme?.lowercased() == "http",
      let expectedPort,
      remotePort == Int(expectedPort),
      let remoteAddress
    else {
      throw NemuNativeHttpAddressPolicyError.blockedDestination
    }

    var literal = remoteAddress
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .trimmingCharacters(in: CharacterSet(charactersIn: "[]"))
    if let zoneIndex = literal.firstIndex(of: "%") {
      literal = String(literal[..<zoneIndex])
    }

    guard literal == "127.0.0.1" || literal == "::1" else {
      throw NemuNativeHttpAddressPolicyError.blockedDestination
    }
  }

  static func isForbiddenHostname(_ hostname: String) -> Bool {
    let normalized = normalizeHostname(hostname)
    if normalized.isEmpty { return true }
    if
      normalized == "localhost" ||
      normalized.hasSuffix(".localhost") ||
      normalized == "local" ||
      normalized.hasSuffix(".local") ||
      normalized == "localdomain" ||
      normalized.hasSuffix(".localdomain") ||
      normalized == "internal" ||
      normalized.hasSuffix(".internal") ||
      normalized == "home.arpa" ||
      normalized.hasSuffix(".home.arpa")
    {
      return true
    }

    return normalized == "metadata" ||
      normalized == "metadata.goog" ||
      normalized.hasSuffix(".metadata.goog") ||
      normalized == "instance-data" ||
      normalized == "instance-data.ec2.internal" ||
      normalized == "metadata.aws.internal" ||
      normalized == "metadata.azure.internal"
  }

  static func isNumericHostname(_ hostname: String) -> Bool {
    if hostname.contains(":") { return true }
    return !hostname.isEmpty && hostname.allSatisfy { character in
      character.isNumber || character == "."
    }
  }

  /// Surge and compatible TUN proxies synthesize DNS answers from RFC 2544's
  /// benchmarking range. These remain non-public everywhere else and are only
  /// eligible when they came from resolving a non-numeric hostname.
  static func isProxySyntheticAddress(_ bytes: [UInt8]) -> Bool {
    return bytes.count == 4 && bytes[0] == 198 && (18...19).contains(Int(bytes[1]))
  }

  static func isPublicAddress(_ bytes: [UInt8]) -> Bool {
    switch bytes.count {
    case 4:
      return isPublicIPv4(bytes, offset: 0)
    case 16:
      return isPublicIPv6(bytes)
    default:
      return false
    }
  }

  private static func isPublicIPv4(_ bytes: [UInt8], offset: Int) -> Bool {
    let first = Int(bytes[offset])
    let second = Int(bytes[offset + 1])
    let third = Int(bytes[offset + 2])
    let fourth = Int(bytes[offset + 3])

    if first == 0 || first == 10 || first == 127 || first >= 224 { return false }
    if first == 100 && (64...127).contains(second) { return false }
    if first == 169 && second == 254 { return false }
    if first == 172 && (16...31).contains(second) { return false }
    if first == 192 && second == 0 && third == 0 && fourth != 9 && fourth != 10 {
      return false
    }
    if first == 192 && second == 0 && third == 2 { return false }
    if first == 192 && second == 88 && third == 99 && fourth != 2 { return false }
    if first == 192 && second == 168 { return false }
    if first == 198 && (18...19).contains(second) { return false }
    if first == 198 && second == 51 && third == 100 { return false }
    if first == 203 && second == 0 && third == 113 { return false }
    return true
  }

  private static func isPublicIPv6(_ bytes: [UInt8]) -> Bool {
    if
      bytes.prefix(10).allSatisfy({ $0 == 0 }) &&
      bytes[10] == 0xff &&
      bytes[11] == 0xff
    {
      return isPublicIPv4(bytes, offset: 12)
    }

    if
      bytes[0] == 0x00 &&
      bytes[1] == 0x64 &&
      bytes[2] == 0xff &&
      bytes[3] == 0x9b &&
      bytes[4..<12].allSatisfy({ $0 == 0 })
    {
      return isPublicIPv4(bytes, offset: 12)
    }

    // Only currently allocated global-unicast space is eligible. This excludes
    // unspecified, loopback, ULA, link/site-local, multicast and reserved space.
    if (bytes[0] & 0xe0) != 0x20 { return false } // 2000::/3

    if bytes[0] == 0x20 && bytes[1] == 0x02 {
      return isPublicIPv4(bytes, offset: 2) // 6to4 embedded IPv4
    }
    if bytes[0] == 0x3f && bytes[1] == 0xff && (bytes[2] & 0xf0) == 0 {
      return false // 3fff::/20 documentation space
    }
    if bytes[0] == 0x20 && bytes[1] == 0x01 {
      let third = bytes[2]
      let fourth = bytes[3]
      if third == 0x00 && fourth == 0x00 { return false } // Teredo special range
      if third == 0x00 && fourth == 0x02 && bytes[4] == 0 && bytes[5] == 0 {
        return false // benchmarking
      }
      if third == 0x0d && fourth == 0xb8 { return false } // documentation
      if third == 0x00 && [UInt8(0x10), 0x20, 0x30].contains(fourth & 0xf0) {
        return false // ORCHID/ORCHIDv2/Drone DETs
      }
    }
    return true
  }

  private static func resolveAddresses(_ hostname: String) throws -> [[UInt8]] {
    var hints = addrinfo()
    hints.ai_family = AF_UNSPEC
    hints.ai_socktype = SOCK_STREAM
    hints.ai_protocol = IPPROTO_TCP

    var head: UnsafeMutablePointer<addrinfo>?
    let status = hostname.withCString { pointer in
      getaddrinfo(pointer, nil, &hints, &head)
    }
    guard status == 0, let first = head else {
      throw NemuNativeHttpAddressPolicyError.unresolvedHost
    }
    defer { freeaddrinfo(first) }

    var output: [[UInt8]] = []
    var current: UnsafeMutablePointer<addrinfo>? = first
    while let pointer = current {
      let info = pointer.pointee
      if info.ai_family == AF_INET, let address = info.ai_addr {
        var value = UnsafeRawPointer(address)
          .assumingMemoryBound(to: sockaddr_in.self)
          .pointee
          .sin_addr
        output.append(withUnsafeBytes(of: &value) { Array($0) })
      } else if info.ai_family == AF_INET6, let address = info.ai_addr {
        var value = UnsafeRawPointer(address)
          .assumingMemoryBound(to: sockaddr_in6.self)
          .pointee
          .sin6_addr
        output.append(withUnsafeBytes(of: &value) { Array($0) })
      }
      current = info.ai_next
    }
    return output
  }

  private static func normalizeHostname(_ hostname: String) -> String {
    return hostname
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .trimmingCharacters(in: CharacterSet(charactersIn: "[]"))
      .trimmingCharacters(in: CharacterSet(charactersIn: "."))
      .lowercased()
  }
}

/// Orders URLSession's completion-handler and metrics callbacks without
/// assuming which one Foundation delivers first. A response waiter is released
/// only after all transaction peers have been accepted; unregistering a task
/// releases retained work fail-closed.
final class NemuNativeHttpPeerValidationGate: @unchecked Sendable {
  private enum State {
    case awaitingMetrics
    case allowed
    case blocked
  }

  private typealias Waiter = @Sendable (Bool) -> Void

  private let lock = NSLock()
  private var states: [Int: State] = [:]
  private var waiters: [Int: [Waiter]] = [:]

  func register(taskIdentifier: Int) {
    lock.lock()
    states[taskIdentifier] = .awaitingMetrics
    lock.unlock()
  }

  func afterValidation(
    taskIdentifier: Int,
    completion: @escaping @Sendable (Bool) -> Void
  ) {
    let immediateResult: Bool?
    lock.lock()
    switch states[taskIdentifier] {
    case .allowed:
      immediateResult = true
    case .blocked, nil:
      immediateResult = false
    case .awaitingMetrics:
      waiters[taskIdentifier, default: []].append(completion)
      immediateResult = nil
    }
    lock.unlock()

    if let immediateResult { completion(immediateResult) }
  }

  func resolve(taskIdentifier: Int, isAllowed: Bool) {
    let callbacks: [Waiter]
    lock.lock()
    guard states[taskIdentifier] != nil else {
      lock.unlock()
      return
    }
    states[taskIdentifier] = isAllowed ? .allowed : .blocked
    callbacks = waiters.removeValue(forKey: taskIdentifier) ?? []
    lock.unlock()

    for callback in callbacks { callback(isAllowed) }
  }

  func unregister(taskIdentifier: Int) {
    let callbacks: [Waiter]
    lock.lock()
    states.removeValue(forKey: taskIdentifier)
    callbacks = waiters.removeValue(forKey: taskIdentifier) ?? []
    lock.unlock()

    for callback in callbacks { callback(false) }
  }
}
